driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -0,0 +1,115 @@
+ * execution context - the server - where the expression is evaluated). Users of
+ * this API must mitigate any risk of applying an expression to some type where
+ * resulting behaviour is not defined by this API (for example, by checking for
+ * null, by ensuring that field types are correctly specified). Likewise, unless
+ * otherwise specified, this API does not define the order of evaluation for all
+ * arguments, and whether all arguments to some expression will be evaluated.
+ *
+ * @see Expressions
+ */
+@Evolving
+public interface Expression {
+
+    /**
+     * Returns logical true if the value of this expression is equal to the
+     * value of the other expression. Otherwise, false.
+     *
+     * @param eq the other expression
+     * @return true if equal, false if not equal
+     */
+    BooleanExpression eq(Expression eq);
+
+    /**
+     * Returns logical true if the value of this expression is not equal to the
+     * value of the other expression. Otherwise, false.
+     *
+     * @param ne the other expression
+     * @return true if equal, false otherwise
+     */
+    BooleanExpression ne(Expression ne);
+
+    /**
+     * Returns logical true if the value of this expression is greater than the
+     * value of the other expression. Otherwise, false.
+     *
+     * @param gt the other expression
+     * @return true if greater than, false otherwise
+     */
+    BooleanExpression gt(Expression gt);
+
+    /**
+     * Returns logical true if the value of this expression is greater than or
+     * equal to the value of the other expression. Otherwise, false.
+     *
+     * @param gte the other expression
+     * @return true if greater than or equal to, false otherwise
+     */
+    BooleanExpression gte(Expression gte);
+
+    /**
+     * Returns logical true if the value of this expression is less than the
+     * value of the other expression. Otherwise, false.
+     *
+     * @param lt the other expression
+     * @return true if less than, false otherwise
+     */
+    BooleanExpression lt(Expression lt);
+
+    /**
+     * Returns logical true if the value of this expression is less than or
+     * equal to the value of the other expression. Otherwise, false.
+     *
+     * @param lte the other expression
+     * @return true if less than or equal to, false otherwise
+     */
+    BooleanExpression lte(Expression lte);
+
+    /**
+     * also checks for nulls
+     * @param other
+     * @return
+     */
+    BooleanExpression isBooleanOr(BooleanExpression other);
+    NumberExpression isNumberOr(NumberExpression other);
+    IntegerExpression isIntegerOr(IntegerExpression other);
+    StringExpression isStringOr(StringExpression other);
+    DateExpression isDateOr(DateExpression other);
+    <T extends Expression> ArrayExpression<T> isArrayOr(ArrayExpression<? extends T> other);
+    <T extends DocumentExpression> T isDocumentOr(T other);
```

Comment:
[optional]

Let's consider whether we should have `DocumentExpression isDocumentOr(DocumentExpression other)` for now here. If we come up with a realization of how to support user subtypes of `DocumentExpression`, and if it will require having a method like the current `isDocumentOr`, then we can add something like `<T extends DocumentExpression> T isUserDocumentOr(T other)`. I am proposing to consider this because it is both possible that either we won't realize how to support user types, or, that it will not need `isDocumentOr` to return a user type because values of user types will be constructed somehow else.

Reply:
It seems that the result might be constructed on the bases of some information in the implementation of the "other" parameter. I think we should leave it for consistency - I think having two methods would be confusing.